File: cpp/creational/abstract_factory.cpp
// 🤖 RoboWorks — Chapter 3: The Product Series
//
// STORY: RoboWorks sells two product series. The Industrial series pairs
// a heavy chassis WITH hydraulic arms; the Domestic series pairs a light
// chassis WITH soft-grip arms. Parts from one series always match — a
// soft-grip arm on a heavy chassis would snap off on day one.
//
// PATTERN: Abstract Factory — provide an interface for creating families
// of related objects without specifying their concrete classes. Use when
// products from one family must be used together.
//
// Build: g++ -std=c++17 abstract_factory.cpp -o abstract_factory

#include <iostream>
#include <memory>

// Abstract products
class Chassis {
public:
    virtual ~Chassis() = default;
    virtual void assemble() const = 0;
};

class Arm {
public:
    virtual ~Arm() = default;
    virtual void attach() const = 0;
};

// Concrete products: the Industrial series
class HeavyChassis : public Chassis {
public:
    void assemble() const override { std::cout << "bolting together a heavy steel chassis\n"; }
};

class HydraulicArm : public Arm {
public:
    void attach() const override { std::cout << "attaching a hydraulic arm\n"; }
};

// Concrete products: the Domestic series
class LightChassis : public Chassis {
public:
    void assemble() const override { std::cout << "clipping together a light alloy chassis\n"; }
};

class SoftGripArm : public Arm {
public:
    void attach() const override { std::cout << "attaching a soft-grip arm\n"; }
};

// Abstract factory: one creation method per product type
class RobotSeriesFactory {
public:
    virtual ~RobotSeriesFactory() = default;
    virtual std::unique_ptr<Chassis> createChassis() const = 0;
    virtual std::unique_ptr<Arm> createArm() const = 0;
};

class IndustrialSeries : public RobotSeriesFactory {
public:
    std::unique_ptr<Chassis> createChassis() const override { return std::make_unique<HeavyChassis>(); }
    std::unique_ptr<Arm> createArm() const override { return std::make_unique<HydraulicArm>(); }
};

class DomesticSeries : public RobotSeriesFactory {
public:
    std::unique_ptr<Chassis> createChassis() const override { return std::make_unique<LightChassis>(); }
    std::unique_ptr<Arm> createArm() const override { return std::make_unique<SoftGripArm>(); }
};

// The production planner only depends on the abstract factory interface.
void produceRobot(const RobotSeriesFactory& series) {
    series.createChassis()->assemble();
    series.createArm()->attach();
}

int main() {
    std::cout << "-- Industrial series --\n";
    produceRobot(IndustrialSeries{});
    std::cout << "-- Domestic series --\n";
    produceRobot(DomesticSeries{});
    return 0;
}

File: python/creational/abstract_factory.py
"""RoboWorks — Chapter 3: The Product Series

STORY: RoboWorks sells two product series. The Industrial series pairs a
heavy chassis WITH hydraulic arms; the Domestic series pairs a light
chassis WITH soft-grip arms. Parts from one series always match — a
soft-grip arm on a heavy chassis would snap off on day one.

PATTERN: Abstract Factory — provide an interface for creating families of
related objects without specifying their concrete classes.

Run: python3 abstract_factory.py
"""

from abc import ABC, abstractmethod


# Abstract products
class Chassis(ABC):
    @abstractmethod
    def assemble(self) -> None: ...


class Arm(ABC):
    @abstractmethod
    def attach(self) -> None: ...


# The Industrial series
class HeavyChassis(Chassis):
    def assemble(self) -> None:
        print("bolting together a heavy steel chassis")


class HydraulicArm(Arm):
    def attach(self) -> None:
        print("attaching a hydraulic arm")


# The Domestic series
class LightChassis(Chassis):
    def assemble(self) -> None:
        print("clipping together a light alloy chassis")


class SoftGripArm(Arm):
    def attach(self) -> None:
        print("attaching a soft-grip arm")


# Abstract factory: one creation method per product type
class RobotSeriesFactory(ABC):
    @abstractmethod
    def create_chassis(self) -> Chassis: ...

    @abstractmethod
    def create_arm(self) -> Arm: ...


class IndustrialSeries(RobotSeriesFactory):
    def create_chassis(self) -> Chassis:
        return HeavyChassis()

    def create_arm(self) -> Arm:
        return HydraulicArm()


class DomesticSeries(RobotSeriesFactory):
    def create_chassis(self) -> Chassis:
        return LightChassis()

    def create_arm(self) -> Arm:
        return SoftGripArm()


def produce_robot(series: RobotSeriesFactory) -> None:
    """The production planner depends only on the abstract factory."""
    series.create_chassis().assemble()
    series.create_arm().attach()


def main():
    print("-- Industrial series --")
    produce_robot(IndustrialSeries())
    print("-- Domestic series --")
    produce_robot(DomesticSeries())


if __name__ == "__main__":
    main()
